python/compare_data_files.py

```diff
@@ -0,0 +1,157 @@
+    old_data: pd.DataFrame,
+    new_data: pd.DataFrame,
+    index_column: str,
+) -> None:
+    unique_indices_from_old = set(old_data[index_column].unique()).difference(
+        set(new_data[index_column].unique())
+    )
+    print("Unique rows in OLD DATA")
+    print(old_data[old_data[index_column].isin(unique_indices_from_old)].to_markdown())
+    unique_indices_from_new = set(new_data[index_column].unique()).difference(
+        set(old_data[index_column].unique())
+    )
+    print("Unique rows in NEW DATA")
+    print(new_data[new_data[index_column].isin(unique_indices_from_new)].to_markdown())
+
+
+def normalize_data(data: pd.DataFrame) -> pd.DataFrame:
+    data = data.infer_objects()
+    for column in data.columns:
+        column_type = data[column].dtype
+        if isinstance(column_type, object) and isinstance(data[column].iloc[0], str):
+            print(f"Normalized string column {column}")
+            data[column] = data[column].str.lower()
+        elif isinstance(column_type, object) and isinstance(data[column].iloc[0], list):
+            data[column] = data[column].apply(tuple)
+            print(f"Normalized list column {column}")
+
+    return data
+
+
+if __name__ == "__main__":
+    print("loading data ...")
+    pd.set_option("display.max_rows", MAX_COMPARISON_ROWS)
+    old_data_raw = load_data_file(filepath=OLD_DATA_PATH)
+    old_data_raw.columns = old_data_raw.columns.str.lower()
+    new_data_raw = load_data_file(filepath=NEW_DATA_PATH)
+    new_data_raw.columns = new_data_raw.columns.str.lower()
+
+    # confirm chosen index column is unique in both datasets
+    for data in [old_data_raw, new_data_raw]:
+        if len(data[INDEX_COLUMN].unique()) != len(data):
+            raise ValueError(f"Column {INDEX_COLUMN} is not unique in a dataset!")
+
+    if NORMALIZE_VALUES:
+        print("normalizing data ...")
+        old_data_raw = normalize_data(old_data_raw)
+        new_data_raw = normalize_data(new_data_raw)
+
+    print("comparing ...")
+    if len(old_data_raw) == len(new_data_raw):
+        compare_data(old_data=old_data_raw, new_data=new_data_raw)
+    else:
+        print("WARNING! Can only compare data of the same length and indices!")
+        print(
+            f"""
+                {len(old_data_raw):,} rows in old data
+                {len(new_data_raw):,} rows in new data
+            """
+        )
+        print("detail differences ...")
+        show_data_with_unique_indices(old_data_raw, new_data_raw, INDEX_COLUMN)
+        common_indices = set(old_data_raw[INDEX_COLUMN].unique()).intersection(
+            set(new_data_raw[INDEX_COLUMN].unique())
+        )
+
+        print("compare common rows ...")
+        print(f"{len(common_indices)} common index values")
+        old_data_to_compare, new_data_to_compare = limit_rows_to_compare(
+            old_data_raw, new_data_raw, INDEX_COLUMN, common_indices
+        )
+        compare_data(old_data_to_compare, new_data_to_compare)
```

Comment:
Just commenting on this file as a whole, I like it as a start to this stuff. I'd imagine there might be some decently mature tools/packages to do some of this too. This sort of stuff should definitely be a module in our future internally published package (if we go that route obvs)

Reply:
no doubt, this is mainly an approach I landed on at my last job that relies on `DataFrame.compare`